lib/Transforms/LowerMemRefCopy.cpp

```diff
@@ -0,0 +1,91 @@
+//===- LowerMemRefCopy.cpp - lower memref.copy pass --------------*- C++
+//-*-===//
+//
+// Copyright 2022 Intel Corporation
+// Part of the IMEX Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This pass lowers memref copyOp to linalg generic operations and enables
+/// simple memref copyOp canonicalization
+///
+//===----------------------------------------------------------------------===//
+
+#include "imex/Transforms/Passes.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LogicalResult.h"
+
+namespace imex {
+#define GEN_PASS_DEF_LOWERMEMREFCOPY
+#include "imex/Transforms/Passes.h.inc"
+} // namespace imex
+
+using namespace mlir;
+using namespace imex;
+
+namespace {
+struct LowerMemRefCopy
+    : public imex::impl::LowerMemRefCopyBase<LowerMemRefCopy> {
+  void runOnOperation() override {
+    auto &domInfo = getAnalysis<DominanceInfo>();
+    auto func = getOperation();
+    // walk through memref.copy ops in the funcion body
+    WalkResult result =
+        func.walk<WalkOrder::PreOrder>([&](memref::CopyOp op) -> WalkResult {
+          if (op->getParentOp() != func)
+            return WalkResult::skip();
+          auto src = op.getSource();
+          auto dst = op.getTarget();
+          // supposed to work on memref type
+          auto srcType = src.getType().cast<MemRefType>();
```

Comment:
`cast` will assert if it isn't possible, you need `dyn_cast`

Reply:
my bad, the td definition ensures that it is memrefType, so eliminate the null check.